extensions/lifecycle/tasks/LifecycleTask.js

```diff
@@ -190,55 +205,131 @@ class LifecycleTask extends BackbeatTask {
+     * @param {array} deleteMarkers - delete markers list
+     * @return {array} merge sorted array
+     */
+    _mergeSortVersionsLists(versions, deleteMarkers) {
+        const sortedList = [];
+        let vIdx = 0;
+        let dmIdx = 0;
+
+        while (vIdx < versions.length || dmIdx < deleteMarkers.length) {
+            if (versions[vIdx] === undefined) {
+                sortedList.push(deleteMarkers[dmIdx++]);
+            } else if (deleteMarkers[dmIdx] === undefined) {
+                sortedList.push(versions[vIdx++]);
+            } else if (versions[vIdx].Key !== deleteMarkers[dmIdx].Key) {
+                // 1. by Key name
+                const isVKeyEarlier = (versions[vIdx].Key <
+                    deleteMarkers[dmIdx].Key);
+                if (isVKeyEarlier) {
+                    sortedList.push(versions[vIdx++]);
+                } else {
+                    sortedList.push(deleteMarkers[dmIdx++]);
+                }
+            } else if (versions[vIdx].LastModified !==
+            deleteMarkers[dmIdx].LastModified) {
+                // 2. by LastModified
+                const isVDateEarlier = (new Date(versions[vIdx].LastModified) >
+                    new Date(deleteMarkers[dmIdx].LastModified));
+                if (isVDateEarlier) {
+                    sortedList.push(versions[vIdx++]);
+                } else {
+                    sortedList.push(deleteMarkers[dmIdx++]);
+                }
+            } else if (versions[vIdx].IsLatest) {
+                // 3. if both above match,
+                //    find if one `IsLatest` and take preference
+                sortedList.push(versions[vIdx++]);
+            } else if (deleteMarkers[dmIdx].IsLatest) {
+                sortedList.push(deleteMarkers[dmIdx++]);
+            } else {
+                // 4. None of above? Check VersionId
+                const isVVersionIdGreater = (versions[vIdx].VersionId >
+                    deleteMarkers[dmIdx].VersionId);
+                if (isVVersionIdGreater) {
+                    sortedList.push(versions[vIdx++]);
+                } else {
+                    sortedList.push(deleteMarkers[dmIdx++]);
+                }
+            }
+        }
+
+        return sortedList;
+    }
+
+    /**
+     * Helper method to apply a staleDate property to each Version and remove
+     * DeleteMarkers from the list
+     * @param {object} bucketDetails - details property from Kafka Bucket entry
+     * @param {string} [bucketDetails.keyMarker] - previous listing key name
+     * @param {string} [bucketDetails.prevDate] - previous listing LastModified
+     * @param {array} list - list of sorted versions and delete markers
+     * @return {array} an updated array of Versions only with applied staleDate
+     */
+    _applyVersionStaleDate(bucketDetails, list) {
+        const appliedList = [];
+
+        for (let i = 0; i < list.length; i++) {
+            // manually check for properties only Versions have and
```

Comment:
I'm not sure you want to ignore delete markers here. If my understanding is correct, delete markers also are candidates for noncurrent version expiration rules. `ExpiredObjectDeleteMarker` only applies to current delete markers with zero non-current versions for the object, which is not the case when delete markers are noncurrent or co-exist with other versions (or other noncurrent delete markers).